chain/chain/src/stateless_validation/chunk_endorsement.rs

```diff
@@ -101,6 +112,60 @@ pub fn validate_chunk_endorsements_in_block(
             tracing::error!(target: "chain", ?endorsement_stats, "Chunk does not have enough stake to be endorsed");
             return Err(Error::InvalidChunkEndorsement);
         }
+
+        // Validate the chunk endorsements bitmap (if present) in the block header against the endorsement signatures in the body.
+        if let Some(endorsements_bitmap) = endorsements_bitmap {
+            for (bit, signature) in
+                endorsements_bitmap.iter(chunk_header.shard_id()).zip(signatures.iter())
+            {
+                if bit != signature.is_some() {
+                    return Err(Error::InvalidChunkEndorsementBitmap(
+                        format!("Chunk endorsement bit in header does not match endorsement in body. shard={}, bit={}, signature={}",
+                            chunk_header.shard_id(), bit, signature.is_some())));
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Validates the [`ChunkEndorsementBitmap`] in the [`BlockHeader`] if it is present, otherwise returns an error.
+/// This function must be called only if ChunkEndorsementInBlockHeader feature is enabled.
+pub fn validate_chunk_endorsements_in_header(
+    epoch_manager: &dyn EpochManagerAdapter,
+    header: &BlockHeader,
+) -> Result<(), Error> {
+    let Some(chunk_endorsements) = header.chunk_endorsements() else {
+        return Err(Error::InvalidChunkEndorsementBitmap(format!(
+            "Expected chunk endorsements bitmap but found None at height {}",
+            header.height()
+        )));
+    };
+    let epoch_id = epoch_manager.get_epoch_id_from_prev_block(header.prev_hash())?;
+    let shard_ids = epoch_manager.get_shard_layout(&epoch_id)?.shard_ids().collect_vec();
+    if chunk_endorsements.num_shards() != shard_ids.len() {
+        return Err(Error::InvalidChunkEndorsementBitmap(
+            format!("Number of shards in bitmap and in epoch do not match: shards in bitmap={}, shards in epoch={}",
+                chunk_endorsements.num_shards(), shard_ids.len())));
+    }
+    for shard_id in shard_ids.into_iter() {
+        let num_validator_assignments = epoch_manager
+            .get_chunk_validator_assignments(&epoch_id, shard_id, header.height())?
+            .len();
+        // 1. Bitmap's length must be equal or greater than the number of assignments in the previous block.
+        if chunk_endorsements.len(shard_id).unwrap() < num_validator_assignments {
```

Comment:
This currently allows very long bitmap with `false` tail. We need precise length check (`ceil(num_assignments / 8) * 8`?) and unit test for that.

Reply:
good point, changing the check and adding tests.

Reply:
actually this requires being able to modify a BlockHeaderV5, so I will do that in follow-up PR (in which I will add mutability to different header versions). stay tuned.